src/lib/tls/asio/asio_async_handshake_op.h

```diff
@@ -18,73 +18,73 @@ namespace Botan {
+
+      void operator()(boost::system::error_code ec, std::size_t bytesTransferred, bool isContinuation = true)
          {
          // process tls packets from socket first
          if(bytesTransferred > 0)
             {
             boost::asio::const_buffer read_buffer {m_core.input_buffer.data(), bytesTransferred};
             try
                {
-               m_channel->received_data(
-                  static_cast<const uint8_t*>(read_buffer.data()),
-                  read_buffer.size());
+               m_stream.native_handle()->received_data(static_cast<const uint8_t*>(read_buffer.data()), read_buffer.size());
                }
             catch(const std::exception&)
                {
                ec = convertException();
-               m_handler(ec);
+               this->invoke(isContinuation, ec);
                return;
                }
             }
 
          // send tls packets
          if(m_core.hasDataToSend())
             {
-            AsyncWriteOperation<AsyncHandshakeOperation<typename std::decay<Handler>::type, StreamLayer, Channel>>
-                  op{std::move(*this), m_core, 0};
-            boost::asio::async_write(m_nextLayer, m_core.sendBuffer(), std::move(op));
+            // \note: we construct `AsyncWriteOperation` with 0 as its last parameter (`plainBytesTransferred`).
+            //        This operation will eventually call `*this` as its own handler, passing the 0 back to this call
+            //        operator. This is necessary because, the check of `bytesTransferred > 0` assumes that
+            //        `bytesTransferred` bytes were just read and are in the cores input_buffer for further processing.
+            AsyncWriteOperation<
+            AsyncHandshakeOperation<typename std::decay<Handler>::type, Stream, Allocator>,
+                                    Stream,
+                                    Allocator>
+                                    op{std::move(*this), m_stream, m_core, 0};
+            boost::asio::async_write(m_stream.next_layer(), m_core.sendBuffer(), std::move(op));
             return;
             }
 
-         if(!m_channel->is_active() && !ec)
+         // we need more tls data from the socket
+         if(!m_stream.native_handle()->is_active() && !ec)
             {
-            // we need more tls data from the socket
-            m_nextLayer.async_read_some(m_core.input_buffer, std::move(*this));
+            m_stream.next_layer().async_read_some(m_core.input_buffer, std::move(*this));
             return;
             }
 
-         if(start)
-            {
-            // don't call the handler directly, similar to io_context.post
-            m_nextLayer.async_read_some(
-               boost::asio::mutable_buffer(m_core.input_buffer.data(), 0), std::move(*this));
-            return;
-            }
-         m_handler(ec);
+         this->invoke(isContinuation, ec);
```

Comment:
So, this is where we will get unnecessary template instantiations. Lets look at the normal calls to read and write. They look like this:
```
boost::asio::async_write(m_stream.next_layer(), m_core.sendBuffer(), std::move(op));
```
and
```
m_stream.next_layer().async_read_some(m_core.input_buffer, std::move(*this));
```

We can leverage the fact that reads and writes with empty buffers complete immediately, "as-if" by `post`. Lets say that the type of `m_core.input_buffer` is `net::const_buffer`. We can thus write:

```
if(isContinuation)
    yield m_stream.next_layer().async_read_some(net::const_buffer, std::move(*this));
```

This accomplishes a `post` without an unnecessary function template instantiation. For a low-level object like the Botan stream, this is very important, because the added instantiation effectively **doubles** the amount of code generated, with no corresponding benefit.

However, the code as written above has some problems. What is `yield`? Well, that's Asio's "stackless fauxroutines" which are macro based, quite powerful, and work without help from the compiler (works in C++11 and even C++03):
https://github.com/boostorg/asio/blob/035775462e5c32d0112175745978f3627e7ee059/include/boost/asio/coroutine.hpp#L22

Why do you need this? Because you need a way, when the zero-byte read completes, for control to return not to the beginning of the function, but to the point just after the zero-byte read, so you can then call `this->invoke_now(...)`. The fauxroutine lets you do that.

There's one other little problem. We don't want to use `bind_handler`, because that would cause `async_read_some` to be instantiated with a different type, doubling the size of the generated code again. But we need to preserve `ec`. The solution is to store the error code as a data member:
````
error_code ec_;
```

Assign `ec_ = ec;` unconditionally as the first line of `operator()`. Then in the call to `invoke_now`, pass `ec_`.